src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs

```diff
@@ -1976,7 +1980,7 @@ protected override BoundNode VisitReturnStatementNoAdjust(BoundReturnStatement n
 
                 if (this.State.ResultIsNotNull == false)
                 {
-                    if ((object)returnType != null && returnType.IsReferenceType && returnType.IsNullable == false)
+                    if ((object)returnType != null && returnType.IsReferenceType && returnType.IsNullable == false && _performStaticNonNullChecks)
```

Comment:
Consider folding this condition with the one above: if (this.State.ResultIsNotNull == false && (object)returnType != null && ...)